diff --git a/pp_sys.c b/pp_sys.c
index 5f8bb0d6ed..90ab46d651 100644
--- a/pp_sys.c
+++ b/pp_sys.c
@@ -2486,7 +2486,7 @@ PP_wrapped(pp_truncate, 2, 0)
     /* XXX Configure probe for the length type of *truncate() needed XXX */
     Off_t len;
 
-#if Off_t_size > IVSIZE
+#if Off_t_size > IVSIZE || defined(__wasi__)
     len = (Off_t)POPn;
 #else
     len = (Off_t)POPi;
@@ -3267,7 +3267,7 @@ PP_wrapped(pp_stat, !(PL_op->op_flags & OPf_REF), 0)
 #else
         PUSHs(newSVpvs_flags("", SVs_TEMP));
 #endif
-#if Off_t_size > IVSIZE
+#if Off_t_size > IVSIZE || defined(__wasi__)
         mPUSHn(PL_statcache.st_size);
 #else
         mPUSHi(PL_statcache.st_size);
@@ -3524,7 +3524,7 @@ PP(pp_ftis)
         dTARGET;
         switch (op_type) {
         case OP_FTSIZE:
-#if Off_t_size > IVSIZE
+#if Off_t_size > IVSIZE || defined(__wasi__)
             sv_setnv(TARG, (NV)PL_statcache.st_size);
 #else
             sv_setiv(TARG, (IV)PL_statcache.st_size);
